cache: add cachecluster resource

sahil-lakhwani opened the pull request:
Signed-off-by: sahil-lakhwani <sahilakhwani@gmail.com>


### Description of your changes

Adds CacheCluster resource

Example: `examples/cache/memcached/cluster.yaml`

### Checklist
<!--
Please run through the below readiness checklist. The first two items are
relevant to every Crossplane pull request.
-->
I have:
- [x] Run `make reviewable` to ensure this PR is ready for review.
- [x] Ensured this PR contains a neat, self documenting set of commits.
- [ ] Updated any relevant [documentation], [examples], or [release notes].
- [ ] Updated the dependencies in [`app.yaml`] to include any new role permissions.

[documentation]: https://github.com/crossplane/crossplane/tree/master/docs
[examples]: https://github.com/crossplane/crossplane/tree/master/cluster/examples
[release notes]: https://github.com/crossplane/crossplane/tree/master/PendingReleaseNotes.md
[`app.yaml`]: https://github.com/crossplane/provider-aws/blob/master/config/stack/manifests/app.yaml

negz commented:
@sahil-lakhwani Could you please move this PR out of draft, have someone from your team review it, and let us know when your internal review process has finished and you're ready for a Crossplane maintainer to review? Thanks!

sahil-lakhwani commented:
Hey @negz , in the planning meeting that of 06/15 - I think you had assigned it to yourself and agreed to review it hence I did not initiate separate review, but I will get it reviewed once internally now.

ajaykangare commented on apis/cache/v1alpha1/cachecluster_types.go:
```diff
@@ -0,0 +1,344 @@
+	runtimev1alpha1 "github.com/crossplane/crossplane-runtime/apis/core/v1alpha1"
+)
+
+// Error strings
+// const (
+// 	errResourceIsNotReplicationGroup = "the managed resource is not a ReplicationGroup"
+// )
```
If this is not required, please remove it.

ajaykangare commented on pkg/controller/cache/cluster/controller.go:
```diff
@@ -0,0 +1,206 @@
+
+	cr.Status.SetConditions(runtimev1alpha1.Creating())
+
+	name := meta.GetExternalName(cr)
+
+	_, err := e.client.CreateCacheClusterRequest(elasticache.NewCreateCacheClusterInput(cr.Spec.ForProvider, name)).Send(ctx)
```
`name` field is used only once, it can be used without a declaration.

ajaykangare commented on apis/cache/v1alpha1/cachecluster_types.go:
```diff
@@ -0,0 +1,344 @@
+	// The password used to access a password protected server.
+	AuthToken *string `json:"authToken,omitempty"`
+
+	// Specifies the strategy to use to update the AUTH token. This parameter must
+	// be specified with the auth-token parameter. Possible values:
+	AuthTokenUpdateStrategy string `json:"authTokenUpdateStrategy,omitempty"`
```
If this is a required field, please remove `omitempty`

ajaykangare commented on apis/cache/v1alpha1/cachecluster_types.go:
```diff
@@ -0,0 +1,344 @@
+	// CacheClusterID is a required field
+	CacheClusterID string `json:"cacheClusterId,omitempty"`
+
+	// A list of cache node IDs to be removed.
+	CacheNodeIdsToRemove []string `json:"cacheNodeIdsToRemove,omitempty"`
```
@sahil-lakhwani  If this is Optional then must be a pointer type.

ajaykangare commented on apis/cache/v1alpha1/cachecluster_types.go:
```diff
@@ -0,0 +1,344 @@
+	// The password used to access a password protected server.
+	AuthToken *string `json:"authToken,omitempty"`
+
+	// Specifies the strategy to use to update the AUTH token. This parameter must
+	// be specified with the auth-token parameter. Possible values:
+	AuthTokenUpdateStrategy string `json:"authTokenUpdateStrategy,omitempty"`
+
+	// The node group (shard) identifier. This parameter is stored as a lowercase
+	// string.
+	// CacheClusterID is a required field
+	CacheClusterID string `json:"cacheClusterId,omitempty"`
```
If this is a required field, please remove `omitempty`

ajaykangare commented:
@sahil-lakhwani  Added a few review comments, kindly address those and other looks good for me.

rahulchheda commented on apis/cache/v1alpha1/cachecluster_types.go:
```diff
@@ -0,0 +1,347 @@
+	// The cache node identifier.
+	CacheNodeID string `json:"cacheNodeId,omitempty"`
+
+	// The current state of this cache node, one of the following values.
```
```suggestion
	// The current state of this cache node, one of the following values: available , creating , deleted , deleting , incompatible-network , modifying , rebooting cluster nodes , restore-failed , or snapshotting .
```

rahulchheda commented on apis/cache/v1alpha1/cachecluster_types.go:
```diff
@@ -0,0 +1,347 @@
+	// +optional
+	CacheNodeIdsToRemove []string `json:"cacheNodeIdsToRemove,omitempty"`
+
+	// The compute and memory capacity of the nodes in the node group (shard).
```
```suggestion
	// The compute and memory capacity of the nodes in the node group (shard).
	// +immutable
```

sahil-lakhwani replied:
`CacheNodeType` can actually be modified.

rahulchheda commented on pkg/clients/elasticache/elasticache.go:
```diff
@@ -398,3 +405,142 @@ func IsSubnetGroupUpToDate(p cachev1alpha1.CacheSubnetGroupParameters, sg elasti
+// NewCreateCacheClusterInput returns Cache Cluster creation input
+func NewCreateCacheClusterInput(p cachev1alpha1.CacheClusterParameters, id string) *elasticache.CreateCacheClusterInput {
+	c := &elasticache.CreateCacheClusterInput{
+		// AZMode: elasticache.AZMode(p.AZMode),
```
can we remove this comment?

rahulchheda commented on pkg/clients/elasticache/elasticache.go:
```diff
@@ -398,3 +405,142 @@ func IsSubnetGroupUpToDate(p cachev1alpha1.CacheSubnetGroupParameters, sg elasti
+	}
+}
+
+// GenerateCluster takes a *CacheClusterParameters and returns *elasticache.CacheCluster.
```
This comment is a little misleading.

sahil-lakhwani commented:
@ajaykangare @rahulchheda I have addressed your comments, let me know if there are any more.

@negz I think this PR is ready for a maintainer to review.

hasheddan commented on apis/cache/v1alpha1/cachecluster_types.go:
```diff
@@ -0,0 +1,341 @@
+
+	// A selector to select a referencer to retrieve the ID of a Security Group
+	// +optional
+	// +immutable
```
I don't think this should be `// +immutable`

hasheddan commented on apis/cache/v1alpha1/cache_subnet_group_types.go:
```diff
@@ -1,12 +1,9 @@
 Licensed under the Apache License, Version 2.0 (the "License");
 you may not use this file except in compliance with the License.
 You may obtain a copy of the License at
-
```
It appears these license headers were unintentionally modified throughout the PR

hasheddan commented on pkg/controller/cache/cluster/controller.go:
```diff
@@ -0,0 +1,201 @@
+	}
+
+	// AWS API rejects modification requests if the state is not `available`
+	if cr.Status.AtProvider.CacheClusterStatus != v1beta1.StatusAvailable {
```
I noticed in the switch block in `Observe` that this is compared to `v1alpha1.StatusAvailable` but here is compared to `v1beta1.StatusAvailable`. I am guessing the former is what you want?

sahil-lakhwani commented:
@hasheddan I have addressed your comments.
